docs/guides/dropout.rst

```diff
@@ -0,0 +1,147 @@
+Applying dropout
+================
+
+This guide provides an overview of how to apply
+`dropout <https://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf>`__
+using :meth:`flax.linen.Dropout`
+
+Dropout is a stochastic regularization technique that randomly removes hidden
+and visible units in a network.
+
+Throughout the guide, you will be able to compare code examples with and without
+Flax ``Dropout``.
+
+.. testsetup::
+
+  import flax.linen as nn
+  import jax.numpy as jnp
+  import jax
+  import optax
+
+Split the PRNG
+**************
+
+Dropout is a random operation, requiring a PRNG state. Flax (like JAX) uses
+`Threefry <https://github.com/google/jax/blob/main/docs/jep/263-prng.md>`__ PRNG
+that is splittable.
+
+**Note:** Recall that JAX has an explicit way of giving you PRNG keys:
+you can fork the main PRNG state (such as ``key = jax.random.PRNGKey(seed=0)``)
+into multiple new PRNG keys with ``key, subkey = jax.random.split(key)``. You
+can refresh your memory in
+`🔪 JAX - The Sharp Bits 🔪 Randomness and PRNG keys <https:/t/jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers>`__.
+
+You split the PRNG into keys, including one for dropout, as follows::
+
+  seed = 0
+  root_key = jax.random.PRNGKey(seed=seed)
+  main_key, params_key, dropout_key = jax.random.split(key=root_key, num=3)
+
+Each of the Flax PRNG streams has a name, such as the ``'params'`` stream for
+initializing parameters, and ``'dropout'`` for dropout.
+
+Defining a model with ``Dropout``
+*********************************
+
+To create a model, subclass :meth:`flax.linen.Module`, and then use
+:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
+
+* If it's set to ``False``, the inputs are scaled by ``1 / (1 - dropout_rate)`` and masked.
+* When it's ``True``, no mask is applied (the dropout is turned off).
+
+Flax provides an implicit way of handling PRNG key streams via Flax
+:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
+It allows the code in Flax ``Module``s (or their sub-``Module``s) to
+"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
+call it.
+
```

Comment:
I suggest removing this paragraph since it contains the same content as the next paragraph, and I think the next one is a bit more clear.

```suggestion
```

Reply:
I think some users may benefit from some repetition. Maybe merging the two paragraphs under one **Note:** may shorten this section:

```
**Note:** Flax provides an implicit way of handling PRNG key streams via Flax
:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
It allows the code in Flax ``Module``s (or their sub-``Module``s) to
"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
call it. Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
to create a key for dropout. You can check out the
`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
```

The internals and `make_rng` is another new guide (idea).

Reply:
Also, I think @andsteing pointed me to the source code, which is awesome. Some Flax users may find this useful too.